### DISCUSSION TITLE
I think the SSO authentication login strategy needs to be optimized

### DISCUSSION THREAD

**Discussion by cdwmhcc** [directus/directus#9273] [Feature Requests]
> I think the SSO authentication login strategy needs to be optimized.

## Situation 1:

The user's email can be obtained using the api interface of providers such as gmail and github, but some other providers cannot obtain the user's email, such as Tencent QQ/Tencent WeChat/Sina Weibo. When using them as a third-party login, it only provides the following user data:

``` JSON
{
"ret":0,
"msg":"",
"nickname":"Peter",
"figureurl":"http://qzapp.qlogo.cn/qzapp/111111/942FEA70050EEAFBD4DCE2C1FC775E56/30",
"gender":"Male",
"openid": "****************"
}
```

At this time, because there is no email, a new user cannot be created after the auth is successful.

For this kind of situation, I think it needs to add a step, when the oauth provider does not provide email, let the user manually bind or register a new user.

## Situation 2:

It is not possible to bind multiple Provider to an account. For example, I have an account: directus@directus.io. I bind this account to github/google/microsoft/facebook. In the future, I can use any one of github/google/microsoft/facebook auth login.

## Situation 3:

When I use a certain vendor to authenticate and log in, I will not be able to log in with a password. I don’t know if this is a bug.

## Situation 4:

After I have created a user, I don't know how to bind the oauth2 provider for this user. The Provider in the app cannot be modified.

It is possible that I did not understand the document, please let me know, thank you.

**Comment by joselcvarela** [comment]
Hello @cdwmhcc,

Here's my thoughts:

1. A recent PR was made in order to fix that, so the email is no longer required for SSO.
https://github.com/directus/directus/pull/9153

2. Here's an issue regarding this: https://github.com/directus/directus/issues/9139
This seems to be related with 1. because that PR should solve this issue.

3. Should this be possible? If you are logging in via an external provider, doesn't make sense to login with a password, right?

4. Recently, the SSO logic was modified in order to support multiple vendors. While in the past we were using `grant` package to handle this, now, thanks to @aidenfoxx, we are using our own implementation. Here's those changes: https://github.com/directus/directus/pull/8660

Those modifications should be reflected on documentation: https://docs.directus.io/configuration/config-options/#authentication

**Reply by cdwmhcc** [reply]
Thank you for your reply.

I have read all of these issue/pr before submitting the issue, but I don't think it can solve the above problem.

1. Only one oauth can be stored (same as point 2)
2. This is not the same question, I mean that a user can bind multiple oauth2 providers, not one of multiple oauth2 providers can be selected.
3. This is not reasonable. Example: If after logging in with oauth, the supplier's account is lost in the future, the user will not be able to log in. I have never seen this logic on a platform that supports oauth third-party login.
4. I don't understand how it relates to the 4th point in my question.

Take medium.com as an example, its process is:

1. You can choose which oauth provider to log in (email registration is listed as one of the choices).

![image](https://user-images.githubusercontent.com/8030134/139638670-0640eeec-23eb-4f0e-9c3e-9d3d567b467f.png)
![image](https://user-images.githubusercontent.com/8030134/139640793-b915fdb6-5684-4dd7-ae3a-8ac1266910b2.png)

2. After using oauth authentication, if the user information returned by the oauth provider contains email, an account will be created, and if there is no email, the user will be guided to create an account manually. Comment: 1
3. After the account is created, the user can bind other oauth providers

![image](https://user-images.githubusercontent.com/8030134/139638589-8eaf8d77-c1bd-4aa3-a86e-43e4655ed921.png)

4. If you bind google and twitter, you can use either of them to log in to medium in the future. You can also log in using email + password.

Currently, if a user of directus is bound to a google provider, the user can only log in with google in the future and cannot choose to bind to other providers.

Hope you can understand what I mean.

**Reply by cdwmhcc** [reply]
Dribbble:

![image](https://user-images.githubusercontent.com/8030134/139641239-60550772-4f67-4bbd-b1cc-448908d5c44d.png)
![image](https://user-images.githubusercontent.com/8030134/139641096-dc940f76-a25f-4db8-91c2-cb72b304f64e.png)
![image](https://user-images.githubusercontent.com/8030134/139641161-f4569f13-746a-4f01-b170-75390b487cbf.png)

**Comment by aidenfoxx** [comment]
@cdwmhcc Holy shit that is some big image spam!

1. This has been solved. You can configure the profile key used for authentication with `AUTH_<provider>_IDENTIFIER_KEY="phone_number"`. Now your OAuth user is authenticated by phone number, not email

2. This was a decision by us to allow secure authentication for other providers (LDAP, Kerberos) where users should only be able to log in with one provider. OAuth isn't the only way people authenticate!!

3. The idea of OAuth is that you shouldn't use a password!

4. You bind the OAuth provider when creating the user. If you want the user to be able to log in with google via email, then you set their "Provider = google" and "External Identifier =myemail@test.com" when creating them

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
